packages/filesystem/src/browser/file-tree/file-tree.ts

```diff
@@ -28,6 +29,35 @@ export class FileTree extends TreeImpl {
     @inject(FileSystem) protected readonly fileSystem: FileSystem;
     @inject(LabelProvider) protected readonly labelProvider: LabelProvider;
 
+    @postConstruct()
+    protected init(): void {
+        this.toDispose.push(
+            this.labelProvider.onDidChange((uri: URI) => this.doUpdateElement(uri))
+        );
+    }
+
+    protected async doUpdateElement(uri: URI): Promise<void> {
+        if (this.root && CompositeTreeNode.is(this.root)) {
+            for (const workspaceRoot of this.root.children) {
+                if (CompositeTreeNode.is(workspaceRoot)) {
+                    const workspaceUri = new URI(workspaceRoot.id);
+                    const isInThisWorkspaceRoot = !!workspaceUri.relative(uri);
+                    if (isInThisWorkspaceRoot) {
+                        const nodeId = this.toNodeId(uri, workspaceRoot);
+                        const node = this.nodes[nodeId];
+                        if (node) {
+                            const mutableNode: MutableTreeNode = node;
```

Comment:
we don't need `MutableTreeNode`, we can do `Mutable<TreeNode>`

Reply:
> it looks good, but won't work properly for diff editors.

Thank you for reviewing so quickly!  I have only started going through the places where labels need updating and diff editors were next on my list.  Thanks for pointing out that this will need a bit more re-work.